### ISSUE TITLE
Add "Parallel map reduce" example

### ISSUE THREAD

**Issue by budziq** [rust-lang-nursery/rust-cookbook#329]
Showcase a basic map-reduce scheme with [rayon](https://crates.io/crates/rayon) crate.

Please discuss the idea here before starting implementation.
cc https://github.com/rust-lang-nursery/rust-cookbook/issues/272

**Comment by j-haj** [comment]
One simple idea I had for this is computing the size of a vector (square root of the sum of the squared components). It's a very simple map-reduce operation, although I could see the argument that it is too mathematical/scientific for a cookbook recipe.

Another idea is to do what seems like the canonical map-reduce example (at least in the Hadoop setting) of word count. We have a vector of sentences, map those sentences to number of words, then reduce by summing the counts.

**Comment by budziq** [comment]
Both seam fine to me :+1:. Are you interested in writing this example?

**Comment by j-haj** [comment]
I'd be happy to take this on as well if you don't mind me working on two in parallel (no pun intended)

**Comment by budziq** [comment]
None taken 😸

**Comment by j-haj** [comment]
I have a toy working example [here](https://github.com/j-haj/rayon-map-reduce/blob/master/map-reduce/src/main.rs). Similar to the other `rayon` example, this is a pretty simple example.

**Comment by budziq** [comment]
Cool! But this example could be replaced by a call to rayons 
[`par_split_whitespace`](https://docs.rs/rayon/0.8.2/rayon/str/trait.ParallelString.html#method.par_split_whitespace) and 
 [`count`](https://docs.rs/rayon/*/rayon/iter/trait.ParallelIterator.html#method.count) without collecting to a vector. 
```rust
paragraph.par_split_whitespace().count();
```

Lets see if we can come up with some example that is not implementable more cleanly via other means. 
Probably we should go with something that first `filter`s then `map`s and finally `reduce`s or `fold`s
At the moment I'm pretty dry in terms of ideas :/ we can brainstorm a little later on if you also come up dry. :+1:

**Comment by j-haj** [comment]
Sorry this has slowed down a bit - it has been a busy week. I will give this some thought over the weekend!

**Comment by budziq** [comment]
no worries!

**Comment by j-haj** [comment]
@budziq I have a more interesting example [here](https://github.com/j-haj/rayon-map-reduce/blob/master/map-reduce/src/main.rs). Admittedly it is still a bit contrived since `Person` isn't really all that useful, but I think the idea is very applicable and the example is simple.

One thing I should point out is that I'm intentionally using `reduce()` over `sum()` here to illustrate the usage of `reduce()`.

**Comment by budziq** [comment]
Nice @j-haj !

Few suggestions:
- Id change the age to some unsigned type
- f64 might be an overkill. How about just f32?
- I'd suggest to use `sum` anyway and maybe show how it can be used with `reduce` too 
- the division part of the code is quite hard to read. I'd split it into two distinct f32 variables.
